File: go/store/kvstore.go
package store

import pb "github.com/psu-csl/replicated-store/go/consensus/multipaxos/comm"

const (
	KeyNotFound string = "key not found"
	Empty              = ""
)

type KVResult struct {
	Ok    bool
	Value string
}

type KVStore interface {
	Get(key string) *string
	Put(key string, value string) bool
	Del(key string) bool
	Execute(cmd *pb.Command) KVResult
}

File: c++/kvstore.cc
#include "json.h"
#include "kvstore.h"
#include "memkvstore.h"
#include "rocksdbstore.h"
#include <glog/logging.h>

namespace kvstore {

std::unique_ptr<KVStore> CreateStore(nlohmann::json const& config) {
  if (config["store"] == "rocksdb") {
    return std::make_unique<RocksDBStore>(config["db_path"]);
  } else {
    return std::make_unique<MemKVStore>();
  }
}

KVResult Execute(multipaxos::Command const& command, KVStore* store) {
  if (command.type() == multipaxos::CommandType::GET) {
    auto r = store->Get(command.key());
    if (!r)
      return KVResult{false, kNotFound};
    return KVResult{true, std::move(*r)};
  }

  if (command.type() == multipaxos::CommandType::PUT) {
    if (store->Put(command.key(), command.value()))
      return KVResult{true, kEmpty};
    return KVResult{false, kPutFailed};
  }

  CHECK(command.type() == multipaxos::CommandType::DEL);

  if (store->Del(command.key()))
    return KVResult{true, kEmpty};
  return KVResult{false, kNotFound};
}

}  // namespace kvstore
